xarray/core/dataset.py

```diff
@@ -1765,50 +1765,61 @@ def _validate_indexers(
         """
         from .dataarray import DataArray
 
-        invalid = [k for k in indexers if k not in self.dims]
+        invalid = indexers.keys() - self.dims.keys()
         if invalid:
             raise ValueError("dimensions %r do not exist" % invalid)
 
         # all indexers should be int, slice, np.ndarrays, or Variable
-        indexers_list = []  # type: List[Tuple[Any, Union[slice, Variable]]]
         for k, v in indexers.items():
-            if isinstance(v, slice):
-                indexers_list.append((k, v))
-                continue
-
-            if isinstance(v, Variable):
-                pass
+            if isinstance(v, (int, slice, Variable)):
+                yield k, v
             elif isinstance(v, DataArray):
-                v = v.variable
+                yield k, v.variable
             elif isinstance(v, tuple):
-                v = as_variable(v)
+                yield k, as_variable(v)
             elif isinstance(v, Dataset):
                 raise TypeError("cannot use a Dataset as an indexer")
             elif isinstance(v, Sequence) and len(v) == 0:
-                v = Variable((k,), np.zeros((0,), dtype="int64"))
+                yield k, np.empty((0,), dtype="int64")
             else:
                 v = np.asarray(v)
 
-                if v.dtype.kind == "U" or v.dtype.kind == "S":
+                if v.dtype.kind in "US":
                     index = self.indexes[k]
                     if isinstance(index, pd.DatetimeIndex):
                         v = v.astype("datetime64[ns]")
                     elif isinstance(index, xr.CFTimeIndex):
                         v = _parse_array_of_cftime_strings(v, index.date_type)
 
-                if v.ndim == 0:
-                    v = Variable((), v)
-                elif v.ndim == 1:
-                    v = Variable((k,), v)
-                else:
+                if v.ndim > 1:
                     raise IndexError(
                         "Unlabeled multi-dimensional array cannot be "
                         "used for indexing: {}".format(k)
                     )
+                yield k, v
```

Comment:
Does it actually meaningfully speed things up to not create a Variable here? I'm OK making things a little more complicated if need be, but it's nice to have slightly simpler logic and Variable objects can be pretty lightweight. Maybe we could use `Variable(..., fastpath=True)`, which skips most of the validation logic?

Reply:
Yes, it's definitely worth it. Check out https://github.com/pydata/xarray/issues/2799#issuecomment-538570946 vs. https://github.com/pydata/xarray/issues/2799#issuecomment-538791352.

This PR reduces the runtime from 1.88s / 10k loops to 1.37s (-27%). Of the original time, 0.25s (13%) is spent on ``Variable.__init__``.

Didn't check fastpath=True. However, consider that Variable.isel *needs to accept any of these inputs anyway*.